### ISSUE TITLE
Gravitational Rule for CA application

### ISSUE THREAD

**Issue by emapco** [emapco/CA-project#9]
Data the function will need: 
- the neighboring cell states as well as our cell of interest state:
  - we might need to template our cell states to be structs that can contain a state property (for compatibility with general library) plus other properties
  - we'll need to store weight (maybe stored as state), current acceleration, current velocity. Position can be stored as the cell's index (cells will need to move)
  - index for each neighboring cell (to compute distances)

Equations will need
- $\vec{F}= M\vec{A}$
- $\vec{F}= \vec{F_1} + \vec{F_2} + \vec{F_3} + \dots$ where each $\vec{F_i}$ is a force vector between our cell of interest and one of it's neighbors. (superposition principle)
- Newton's Gravitational force: $\vec{F} = G\frac{m_1*m_2}{|r|^2} * \hat{r}$ where $r$ is the radial distance and $\hat{r} = \frac{\vec{r}}{|r|}$
  - maybe we can set G to 1 to simplify the model 
- kinematic equations:
  - acceleration velocity, position
  - https://physics.stackexchange.com/questions/346229/equations-of-motion-in-vector-form

Some ideas/concerns:
- We can use the above equations to update cell positions
  - We need some kind of tie-breaker or special handling when two cells move to the same cell in the same step
- I'll look into generating a neighbors vector/matrix/tensor so that we can pass it to the custom rule function.
  - This will improve some of the step function logic I have written. Currently the code calculates both the sum of cell states and votes even though only one or neither is needed.
  - I might "flatten" the neighbors and additionally store their original indices (might not be needed; might be able to reverse the logic to decipher their original indices based on their flatten array position and tensor dimensions.

I'm probably missing some things so please provide input.

**Comment by emapco** [comment]
pseudocode (mixture of python (simple syntax) and c++ (types)) of a possible implementation:
```
struct cell
{
    int state; // mass or velocity
    double velocity or mass;
}

galaxy_formatiion_rule(int *cell_index, const int index_size,
                 cell *neighborhood_cells, const int neighborhood_size,
                 cell &new_cell_state):
    
    int total_force[3] // sum of all forces for each coordinate
    int neighbor_index[3] // x, y, z coordinates 
    cell cell_of_interest; // struct

    for index, neighbor_cell in neighborhood_cells:

        // get index from flattened neighborhood_cell array
        get_periodic_moore_neighbor_index(rank, radius, index, *neighbor_index)

        // get cell struct from neighborhood
        get_periodic_middle_cell_from_neighborhood(cell_of_interest)

        int force[3]
        compute_gravitational_force(cell_of_interest, neighbor_cell, neighbor_index)

        // add force to total
        total_force += force

    compute_accel(total_force, cell_of_interest.mass)
    compute_velocity(accel, cell_of_interest)
    compute_new_position(cell_index)

    // will need to handle cases where another cell occupies the computed position
    // or when a cell is in path to new position
    set_new_position(new_cell_state)

    return new_cell_state


compute_gravitational_force(cell, cell_of_interest, cell neighbor_cell, neighbor_index):
    cell_of_interest_index = [0,0,0]
    r = compute_distance_vector(cell_of_interest_index, neighbor_index)
    r_norm = compute_distance_norm(cell_of_interest_index, neighbor_index)
    r_hat = r / r_norm

    force_scalar = mass1 * mass 2 / r_norm
    force_vector = force_scalar * r_hat

    return force_vector


compute_accel(total_force, mass):
    total_force/mass 

compute_velocity(accel, timestep):
    accel * timestep

compute_new_position(velocity, cell_of_interest, cell_index, timestep)
     delta_r = (velocity - cell_of_interst.velocity) * timestep
     cell_index += delta_r
```

**Comment by emapco** [comment]
We decided to allow user to specify merging type of collision and stopping collisions.  Resolved with #18.
